pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -72,22 +72,54 @@ def __init__(
     ) -> None:
         super().__init__(parallel_devices=parallel_devices, cluster_environment=cluster_environment)
         self.interactive_ddp_procs = []
-        self.num_nodes = num_nodes
+        self._num_nodes = num_nodes
         self.sync_batchnorm = sync_batchnorm
         self.dist = LightningDistributed()
         self._ddp_kwargs = kwargs
         self._has_spawned_children = False
         self.task_idx = None
-        self.num_processes = len(parallel_devices) if parallel_devices is not None else parallel_devices
         self._ddp_comm_state = ddp_comm_state
         self._ddp_comm_hook = ddp_comm_hook
         self._ddp_comm_wrapper = ddp_comm_wrapper
+        # world ranks is related to num_nodes, cluster_environment and parallel_devices
+        # when resetting these parameters, need to reset world ranks
         self.set_world_ranks()
 
     @property
     def root_device(self):
         return self.parallel_devices[self.local_rank]
 
+    @property
+    def num_nodes(self):
+        return self._num_nodes
+
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int):
+        self._num_nodes = num_nodes
+        self.set_world_ranks()
```

Comment:
why call set world ranks in the num nodes setter?

Reply:
it is needed as it is used to set world size

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp.py#L219